File: Python3/2516. Take K of Each Character From Left and Right.py
class Solution:
    def takeCharacters(self, s: str, k: int) -> int:
        count = [0] * 3
        for c in s:
            count[ord(c) - ord('a')] += 1

        # Special case
        for c in count:
            if c < k:
                return -1

        ans = l = 0
        for r, c in enumerate(s):
            count[ord(c) - ord('a')] -= 1
            while l <= r and count[ord(c) - ord('a')] < k:
                count[ord(s[l]) - ord('a')] += 1
                l += 1
            
            if count[ord(c) - ord('a')] >= k:
                ans = max(ans, r - l + 1)

        return len(s) - ans

File: Java/2516. Take K of Each Character From Left and Right.java
class Solution {
    public int takeCharacters(String s, int k) {
        int[] count = new int[3];
        for (char c: s.toCharArray()) {
            count[c - 'a']++;
        }

        for (int c: count) {
            if (c < k) {
                return -1;
            }
        }

        int ans = 0, l = 0;
        for (int r = 0; r < s.length(); r++) {
            char c = s.charAt(r);
            count[c - 'a']--;
            while (l <= r && count[c - 'a'] < k) {
                count[s.charAt(l++) - 'a']++;
            }
            if (count[c - 'a'] >= k) {
                ans = Math.max(ans, r - l + 1);
            }
        }

        return s.length() - ans;
    }
}